src/TaskLineRenderer.ts

```diff
@@ -31,6 +31,55 @@ export const LayoutClasses: { [c in TaskLayoutComponent]: string } = {
     blockLink: '',
 };
 
+type AttributeValueCalculator = (component: TaskLayoutComponent, task: Task) => string | null;
+
+export class AppleSauce {
+    className: string;
+    dataAtrributeName: string;
+    attributeValueCalculator: AttributeValueCalculator;
+
+    constructor(className: string, dataAttributeName: string, attributeValueCalculator: AttributeValueCalculator) {
+        this.className = className;
+        this.dataAtrributeName = dataAttributeName;
+        this.attributeValueCalculator = attributeValueCalculator;
+    }
+}
+
+const dateDataAttributeCalculator: AttributeValueCalculator = (component: TaskLayoutComponent, task: Task) => {
+    const date = task[component];
+    if (date instanceof window.moment) {
+        const attributeValue = dateToAttribute(date);
+        if (attributeValue) {
+            return attributeValue;
+        }
+    }
+
+    return null;
+};
+
+const noDataAttribute: AttributeValueCalculator = () => {
+    return null;
+};
+
+const appleSauceDictionary: { [name: string]: AppleSauce } = {
+    createdDate: new AppleSauce('task-created', 'taskCreated', dateDataAttributeCalculator),
+    dueDate: new AppleSauce('task-due', 'taskDue', dateDataAttributeCalculator),
+    startDate: new AppleSauce('task-start', 'taskStart', dateDataAttributeCalculator),
+    scheduledDate: new AppleSauce('task-scheduled', 'taskScheduled', dateDataAttributeCalculator),
+    doneDate: new AppleSauce('task-done', 'taskDone', dateDataAttributeCalculator),
+
+    // TODO the dataAttributeName here being empty is untested - setting it to 'foobar' doesn't change the outcome of the tests.
+    // This was the case before the latest refactoring as well.
```

Comment:
Sorry I wasn't clear about this...
If the test didn't fail, the idea is to go ahead and write a test based on the current implementation.

Not to do the refactoring anyway!

Reply:
Any new test that's added now would be first written and tested against the code on `main`, and then pasted in to the branch, to confirm that the change on this branch is safe.

Reply:
Well, it is a bit tricky to test the absence, but I tried to.

I also cherry-picked the test commit on the main and successfully broke the code there as well. Obviously the way to break it was different there.

I'm still not sure how to test the TOTAL absence of the `blockLink`. Do you have any idea?

Reply:
> I'm still not sure how to test the TOTAL absence of the blockLink. Do you have any idea?

I don't know the details of the code, but I would try to write a test that:

1. renders a task with one or two fields and NO blockLink
2. renders another task that differs in that it only also HAS a block link...

And confirm that the results are identical.... (without specifying what the actual values are - just that rendering a blocklink does not change the result)

Reply:
Also, note the previous open PR: #2368

Reply:
> Also, note the previous open PR: #2368

I think this is a very nice case demonstrating how hard it is now to change the CSS & data attribute related things. I can help with that PR after we merge this one (guess it is better to make the change easy first).

Reply:
Actually if that PR is merged first, then the test for the blockId shall be different and there is no need to write a new one =)